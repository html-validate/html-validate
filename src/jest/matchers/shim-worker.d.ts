declare module "*?worker&url" {
	const value: string;
	export default value;
}
