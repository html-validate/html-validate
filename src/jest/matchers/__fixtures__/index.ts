export { reportOk, reportOkAsync } from "./report-ok";
export { reportError, reportErrorAsync } from "./report-error";
export { reportMultipleErrors, reportMultipleErrorsAsync } from "./report-multiple-errors";
