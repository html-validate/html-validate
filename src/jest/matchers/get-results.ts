import { FileSystemConfigLoader } from "../../config/loaders/file-system";
import { type Report, type Result } from "../../reporter";
import HtmlValidate from "../../htmlvalidate";

/**
 * @internal
 */
export function getResults(filename: string, value: Report | string): Result[] {
	if (typeof value === "string") {
		const loader = new FileSystemConfigLoader({
			extends: ["html-validate:recommended"],
		});
		const htmlvalidate = new HtmlValidate(loader);
		const report = htmlvalidate.validateString(value, filename, {
			rules: {
				"void-style": "off",
			},
		});
		return report.results.map((it) => {
			return { ...it, filePath: "inline" };
		});
	} else {
		return value.results;
	}
}
