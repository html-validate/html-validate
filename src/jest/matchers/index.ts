export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
