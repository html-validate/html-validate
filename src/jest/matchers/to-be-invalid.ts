import { Report } from "../../reporter";
import { diverge } from "../utils";

function toBeInvalid(report: Report): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () => "Result should not contain error",
		};
	}
}

export default diverge(toBeInvalid);
