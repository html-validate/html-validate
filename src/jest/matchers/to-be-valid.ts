import { Report } from "../../reporter";

export async function toBeValid(
	this: jest.MatcherUtils,
	actual: Report | Promise<Report>
): Promise<jest.CustomMatcherResult> {
	const report = await actual;
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () => "Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () => `Result should be valid but had error "${firstError.message}"`,
		};
	}
}
