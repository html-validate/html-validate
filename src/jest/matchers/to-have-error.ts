import { type Message, type Report } from "../../reporter";
import { type MatcherContext, type MatcherResult, diff, diverge } from "../utils";
import { flattenMessages } from "../utils/flatten-messages";

function toHaveErrorImpl(
	jest: MatcherContext,
	actual: Report,
	expected: Partial<Message>
): MatcherResult {
	const flattened = flattenMessages(actual);
	const matcher = [expect.objectContaining(expected)];
	const pass = jest.equals(flattened, matcher);
	const diffString = diff(matcher, flattened, { expand: jest.expand });
	const hint = jest.utils.matcherHint(".toHaveError");
	const prettyExpected = jest.utils.printExpected(matcher);
	const prettyReceived = jest.utils.printReceived(flattened);
	const resultMessage = (): string => {
		return [
			hint,
			"",
			"Expected error to equal:",
			`  ${prettyExpected}`,
			"Received:",
			`  ${prettyReceived}`,
			/* istanbul ignore next */ diffString ? `\nDifference:\n\n${diffString}` : "",
		].join("\n");
	};
	return { pass, message: resultMessage };
}

function toHaveError(this: MatcherContext, actual: Report, error: Partial<Message>): MatcherResult;
function toHaveError(
	this: MatcherContext,
	actual: Report,
	ruleId: string,
	message: string,
	context?: any
): MatcherResult;
function toHaveError(
	this: MatcherContext,
	actual: Report,
	arg1: string | Partial<Message>,
	arg2?: string,
	arg3?: any
): MatcherResult {
	if (typeof arg1 === "string") {
		const expected: Partial<Message> = {
			ruleId: arg1,
			message: arg2,
		};
		if (arg3) {
			/* eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- this is supposed to accept anything */
			expected.context = arg3;
		}
		return toHaveErrorImpl(this, actual, expected);
	} else {
		return toHaveErrorImpl(this, actual, arg1);
	}
}

export default diverge(toHaveError);
