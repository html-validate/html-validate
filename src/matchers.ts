/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/ban-ts-comment, prefer-template, sonarjs/no-duplicate-string */

import diff from "jest-diff";
import deepmerge from "deepmerge";
import { TokenType } from "./lexer";
import { Message, Report, Result } from "./reporter";
import { ConfigData } from "./config";
import HtmlValidate from "./htmlvalidate";

interface TokenMatcher {
	type: TokenType;
	location?: any;
	data?: any;
}

declare global {
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toBeToken(expected: TokenMatcher): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(
				errors: Array<[string, string] | Record<string, unknown>>
			): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 */
			toHTMLValidate(config?: ConfigData): R;
		}
	}
}

/**
 * Takes all messages from all files and flattens to a single array.
 */
function flattenMessages(report: Report): Message[] {
	return report.results.reduce((aggregated: Message[], result: Result) => {
		return aggregated.concat(result.messages);
	}, []);
}

function toBeValid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () =>
				`Result should be successful but had error "${firstError.message}"`,
		};
	}
}

function toBeInvalid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	}
}

function toHaveError(
	this: jest.MatcherUtils,
	report: Report,
	ruleId: any,
	message: any,
	context?: any
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const expected: any = { ruleId, message };
	if (context) {
		expected.context = context;
	}

	const matcher = [expect.objectContaining(expected)];
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveError") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHaveErrors(
	this: jest.MatcherUtils,
	report: Report,
	errors: Array<[string, string] | Record<string, unknown>>
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const matcher = errors.map((entry) => {
		if (Array.isArray(entry)) {
			const [ruleId, message] = entry;
			return expect.objectContaining({ ruleId, message });
		} else {
			return expect.objectContaining(entry);
		}
	});
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveErrors") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHTMLValidate(
	this: jest.MatcherUtils,
	// @ts-ignore DOM library not available
	actual: string | HTMLElement,
	userConfig?: ConfigData,
	filename?: string
): jest.CustomMatcherResult {
	// @ts-ignore DOM library not available
	if (actual instanceof HTMLElement) {
		actual = actual.outerHTML;
	}

	const defaultConfig = {
		rules: {
			/* jsdom normalizes style so disabling rule when using this matcher or it
			 * gets quite noisy when configured with self-closing */
			"void-style": "off",
		},
	};
	const config = deepmerge(defaultConfig, userConfig || {});

	const actualFilename = filename || this.testPath;
	const htmlvalidate = new HtmlValidate();
	const report = htmlvalidate.validateString(actual, actualFilename, config);
	const pass = report.valid;
	if (pass) {
		return { pass, message: () => "HTML is valid when an error was expected" };
	} else {
		const errors = report.results[0].messages.map(
			(message) => `  ${message.message} [${message.ruleId}]`
		);
		return {
			pass,
			message: () =>
				["Expected HTML to be valid but had the following errors:", ""]
					.concat(errors)
					.join("\n"),
		};
	}
}

function toBeToken(
	this: jest.MatcherUtils,
	actual: any,
	expected: any
): jest.CustomMatcherResult {
	const token = actual.value;

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (token.type) {
		token.type = TokenType[token.type];
	}

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (expected.type) {
		expected.type = TokenType[expected.type];
	}

	const matcher = expect.objectContaining(expected);
	const pass = this.equals(token, matcher);
	const diffString = diff(matcher, token, { expand: this.expand });
	const message = (): string =>
		this.utils.matcherHint(".toBeToken") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(token)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message };
}

expect.extend({
	toBeValid,
	toBeInvalid,
	toHaveError,
	toHaveErrors,
	toHTMLValidate,
	toBeToken,
});
