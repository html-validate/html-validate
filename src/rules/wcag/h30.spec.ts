import HtmlValidate from "../../htmlvalidate";
import "../../matchers";

describe("wcag/h30", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			rules: { "wcag/h30": "error" },
		});
	});

	it("should not report when link has text", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString("<a>lorem ipsum</a>");
		expect(report).toBeValid();
	});

	it("should not report when link has image with alt-text", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString('<a><img alt="lorem ipsum"></a>');
		expect(report).toBeValid();
	});

	it("should not report when link has aria-label", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString('<a aria-label="lorem ipsum"></a>');
		expect(report).toBeValid();
	});

	it("should not report when descendant has aria-label", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateString('<a><span aria-label="lorem ipsum"></span></a>');
		expect(report).toBeValid();
	});

	it("should report error when link is missing text", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<a></a>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("wcag/h30", "Anchor link must have a text describing its purpose");
	});

	it("should report error when link is missing text and image alt", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString("<a><img></a>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("wcag/h30", "Anchor link must have a text describing its purpose");
	});

	it("should report error when link is missing text and image has empty alt", () => {
		expect.assertions(2);
		const report = htmlvalidate.validateString('<a><img alt=""></a>');
		expect(report).toBeInvalid();
		expect(report).toHaveError("wcag/h30", "Anchor link must have a text describing its purpose");
	});

	it("smoketest", () => {
		expect.assertions(1);
		const report = htmlvalidate.validateFile("test-files/rules/wcag/h30.html");
		expect(report.results).toMatchSnapshot();
	});

	it("should contain documentation", () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "wcag/h30": "error" },
		});
		expect(htmlvalidate.getRuleDocumentation("wcag/h30")).toMatchSnapshot();
	});
});
