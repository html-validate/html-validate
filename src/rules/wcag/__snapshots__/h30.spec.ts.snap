// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wcag/h30 should contain documentation 1`] = `
Object {
  "description": "WCAG 2.1 requires each \`<a>\` anchor link to have a text describing the purpose of the link using either plain text or an \`<img>\` with the \`alt\` attribute set.",
  "url": "https://html-validate.org/rules/h30.html",
}
`;

exports[`wcag/h30 smoketest 1`] = `
Array [
  Object {
    "errorCount": 4,
    "filePath": "test-files/rules/wcag/h30.html",
    "messages": Array [
      Object {
        "column": 2,
        "context": undefined,
        "line": 7,
        "message": "Anchor link must have a text describing its purpose",
        "offset": 128,
        "ruleId": "wcag/h30",
        "selector": "a:nth-child(4)",
        "severity": 2,
        "size": 1,
      },
      Object {
        "column": 2,
        "context": undefined,
        "line": 8,
        "message": "Anchor link must have a text describing its purpose",
        "offset": 136,
        "ruleId": "wcag/h30",
        "selector": "a:nth-child(5)",
        "severity": 2,
        "size": 1,
      },
      Object {
        "column": 2,
        "context": undefined,
        "line": 9,
        "message": "Anchor link must have a text describing its purpose",
        "offset": 149,
        "ruleId": "wcag/h30",
        "selector": "a:nth-child(6)",
        "severity": 2,
        "size": 1,
      },
      Object {
        "column": 2,
        "context": undefined,
        "line": 10,
        "message": "Anchor link must have a text describing its purpose",
        "offset": 166,
        "ruleId": "wcag/h30",
        "selector": "a:nth-child(7)",
        "severity": 2,
        "size": 1,
      },
    ],
    "source": "<!-- valid cases -->
<a>lorem ipsum</a>
<a><span>lorem ipsum</span></a>
<a><img alt=\\"lorem ipsum\\"></a>

<!-- invalid cases -->
<a></a>
<a><img></a>
<a><img alt></a>
<a><img alt=\\"\\"></a>
",
    "warningCount": 0,
  },
]
`;
